File: 2Arrays and Strings/Code/Cpp/2-2Palindrome.cpp
#include <iostream>
using namespace std;

bool checkIfPalindrome(string s) {
    int left = 0;
    int right = s.size() - 1;
    
    while (left < right) {
        if (s[left] != s[right]) {
            return false;
        }
        left++;
        right--;
    }
    
    return true;
}

int main() {
    /* Odd length strings */
    string s = "racecar"; // palindrome
    printf("%d\n", checkIfPalindrome(s));  // 1
    
    s = "aslkd"; // not palindrome
    printf("%d\n", checkIfPalindrome(s)); // 0
    
    /* Even length strings */
    s = "basdab"; // not palindrome
    printf("%d\n", checkIfPalindrome(s)); // 0

    s = "abassaba"; // palindrome
    printf("%d\n", checkIfPalindrome(s));  // 1
    
    return 0;
}

File: 2Arrays and Strings/Code/Java/2-2Palindrome.java
class Solution {
    public boolean checkIfPalindrome(String s) {
        int left = 0;
        int right = s.length() - 1;

        while (left < right) {
            if (s.charAt(left) != s.charAt(right)) {
                return false;
            }

            left++;
            right--;
        }

        return true;
    }
}

class Main {
    public static void main(String[] args) {
        Solution solution = new Solution();
		    /* Odd length strings */
      
        System.out.println(solution.checkIfPalindrome("racecar")); // prints true
		    System.out.println(solution.checkIfPalindrome("aslkd")); // prints false
		
		    /* Even length strings */
		    System.out.println(solution.checkIfPalindrome("basdab")); // prints false
		    System.out.println(solution.checkIfPalindrome("abassaba")); // prints true
    }
};

/* Compilation steps */
// javac 2-2Palindrome.java
// java Main
